src/nix/formatter.cc

```diff
@@ -0,0 +1,174 @@
+    {
+        return Strings{"formatter." + settings.thisSystem.get()};
+    }
+
+    Strings getDefaultFlakeAttrPathPrefixes() override
+    {
+        return Strings{};
+    }
+
+    void run(ref<Store> store) override
+    {
+        auto evalState = getEvalState();
+        auto evalStore = getEvalStore();
+
+        auto installable_ = parseInstallable(store, ".");
+        auto & installable = InstallableValue::require(*installable_);
+        auto app = installable.toApp(*evalState).resolve(evalStore, store);
+
+        Strings programArgs{app.program};
+
+        // Propagate arguments from the CLI
+        for (auto & i : args) {
+            programArgs.push_back(i);
+        }
+
+        // Release our references to eval caches to ensure they are persisted to disk, because
+        // we are about to exec out of this process without running C++ destructors.
+        evalState->evalCaches.clear();
+
+        execProgramInStore(store, UseLookupPath::DontUse, app.program, programArgs);
+    };
+};
+
+static auto rFormatterRun = registerCommand2<CmdFormatterRun>({"formatter", "run"});
+
+struct CmdFormatterBuild : SourceExprCommand
+{
+    Path outLink = "result";
+
+    CmdFormatterBuild()
+    {
+        addFlag({
+            .longName = "out-link",
+            .shortName = 'o',
+            .description = "Use *path* as prefix for the symlink to the build result. It defaults to `result`.",
+            .labels = {"path"},
+            .handler = {&outLink},
+            .completer = completePath,
+        });
+
+        addFlag({
+            .longName = "no-link",
+            .description = "Do not create symlink to the build results.",
+            .handler = {&outLink, Path("")},
+        });
+    }
+
+    std::string description() override
+    {
+        return "build the current flake's formatter";
+    }
+
+    std::string doc() override
+    {
+        return
+#include "formatter-build.md"
+            ;
+    }
+
+    Category category() override
+    {
+        return catSecondary;
+    }
+
+    Strings getDefaultFlakeAttrPaths() override
+    {
+        return Strings{"formatter." + settings.thisSystem.get()};
+    }
+
+    Strings getDefaultFlakeAttrPathPrefixes() override
+    {
+        return Strings{};
+    }
+
+    void run(ref<Store> store) override
+    {
+        auto evalState = getEvalState();
+        auto evalStore = getEvalStore();
+
+        auto installable_ = parseInstallable(store, ".");
+        auto & installable = InstallableValue::require(*installable_);
+        auto unresolvedApp = installable.toApp(*evalState);
+        auto app = unresolvedApp.resolve(evalStore, store);
+
+        Installables installableContext;
+        for (auto & ctxElt : unresolvedApp.unresolved.context)
+            installableContext.push_back(make_ref<InstallableDerivedPath>(store, DerivedPath{ctxElt}));
+        auto buildables = Installable::build(evalStore, store, Realise::Outputs, installableContext);
```

Comment:
I'm not sure if there's a less awkward way of doing this. I basically copied this logic from here: https://github.com/jfly/nix/blob/7706cff51871449d27b96e528b784a4eb630031d/src/nix/app.cc#L137-L143

Reply:
Instead of copying, I think you could factor out a method.

Reply:
Sounds good! I've pulled this out into a shared `UnresolvedApp::build` method.